### PR TITLE
XL/erasure-read: refactor erasure read

### PR DESCRIPTION
**Pull request by krishnasrinivas**

### CODE REVIEW

File: erasure-readfile.go [hunk 1]
```diff
@@ -188,39 +193,31 @@ func erasureReadFile(writer io.Writer, disks []StorageAPI, volume string, path s
...
+	curChunkSize := chunkSize
+	curBlockSize := blockSize
 
 	// For each block, read chunk from each disk. If we are able to read all the data disks then we don't
 	// need to read parity disks. If one of the data disk is missing we need to read DataBlocks+1 number
 	// of disks. Once read, we Reconstruct() missing data if needed and write it to the given writer.
-	for block := startBlock; bytesWritten < length; block++ {
+	for block := startBlock; block <= endBlock; block++ {
 		// Each element of enBlocks holds curChunkSize'd amount of data read from its corresponding disk.
 		enBlocks := make([][]byte, len(disks))
```
Comment (harshavardhana) [inline erasure-readfile.go:214]: Bring back a comment here.. explain what the `if condition` means.

File: erasure-readfile_test.go [hunk 2]
```diff
@@ -18,8 +18,9 @@ package main
 
 import (
```
Comment (harshavardhana) [inline erasure-readfile_test.go:19]: This file has a license header which says '2015' remove that, keep only 2016.

File: erasure-utils.go [hunk 3]
```diff
@@ -136,20 +136,12 @@ func writeDataBlocks(dst io.Writer, enBlocks [][]byte, dataBlocks int, outOffset
...
-// calculate the blockSize based on input length and total number of
-// data blocks.
-func getEncodedBlockLen(inputLen int64, dataBlocks int) (curEncBlockSize int64) {
-	curEncBlockSize = (inputLen + int64(dataBlocks) - 1) / int64(dataBlocks)
-	return curEncBlockSize
+// chunkSize is roughly BlockSize/DataBlocks.
+// chunkSize is calculated such that chunkSize*DataBlocks accommodates BlockSize bytes.
+// So chunkSize*DataBlocks can be slightly larger than BlockSize if BlockSize is not divisible by
+// DataBlocks. The extra space will have 0-padding.
+func getChunkSize(blockSize int64, dataBlocks int) int64 {
```
Comment (harshavardhana) [inline erasure-utils.go:143]: Can you a testcase for this? catering both the cases?

File: erasure-readfile_test.go [hunk 4]
```diff
@@ -359,3 +362,65 @@ func TestErasureReadFileOffsetLength(t *testing.T) {
...
+	if size != length {
+		t.Errorf("erasureCreateFile returned %d, expected %d", size, length)
+	}
+
+	// To generate random offset/length.
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	buf := &bytes.Buffer{}
+
+	for i := 0; i < iterations; i++ {
```
Comment (harshavardhana) [inline erasure-readfile_test.go:411]: Add a comment here.

File: erasure-readfile_test.go [hunk 5]
```diff
@@ -359,3 +362,65 @@ func TestErasureReadFileOffsetLength(t *testing.T) {
...
 }
+
+// Test erasureReadFile with random offset and lengths.
+// This test is t.Skip()ed as it a long time to run, hence should be run
+// explicitly after commenting out t.Skip()
+func TestErasureReadFileRandomOffsetLength(t *testing.T) {
+	// Comment the following line to run this test.
+	t.SkipNow()
+	// Initialize environment needed for the test.
+	dataBlocks := 7
```
Comment (harshavardhana) [inline erasure-readfile_test.go:373]: Do this for various dataBlocks one for each  i.e 2x2 (100), 3x3(100), 4x4(100), 5x5(100), 6x6(100), 7x7(100), 8x8(100) -- 100 is iterations here.

File: erasure-readfile_test.go [hunk 6]
```diff
@@ -359,3 +362,65 @@ func TestErasureReadFileOffsetLength(t *testing.T) {
...
+	setup, err := newErasureTestSetup(dataBlocks, parityBlocks, blockSize)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	defer setup.Remove()
+
+	disks := setup.disks
+
+	// Prepare a slice of 1MB with random data.
```
Comment (harshavardhana) [inline erasure-readfile_test.go]: Prepare a slice of '5MB'

File: erasure-readfile.go (hunk 1 above)
Comment (krishnasrinivas) [inline erasure-readfile.go:214]: previous comment was not clear, i have added a new comment with an example to be more clear. (will push it)

File: erasure-readfile_test.go (hunk 4 above)
Comment (krishnasrinivas) [inline erasure-readfile_test.go:411]: comment here will not be helpful as the variable names and code flow are self explanatory

File: erasure-readfile_test.go (hunk 5 above)
Comment (krishnasrinivas) [inline erasure-readfile_test.go:373]: Correct, i have this on mind, it will be an exhaustive test coverage, will do this as a separate commit.

Comment (krishnasrinivas) [inline erasure-readfile_test.go:373]: not just with with different data/parity blocks, but also different blockSize. it will be a thorough checking for erasureCreateFile and erasureReadFile

**Comment by harshavardhana** [comment]
> comment here will not be helpful as the variable names and code flow are self explanatory

It is not about that a 'gist' is necessary before each `for loop` why are we iterating and what does the entire 'for block' does is important to be written.

**Comment by hackintoshrao** [comment]
@krishnasrinivas : After the changes can you run this script with hash of your commit and the previous commit to see the performance changes  https://github.com/minio/minio/blob/master/dist/benchmark/benchcmp.sh

**Comment by krishnasrinivas** [comment]
> It is not about that a 'gist' is necessary before each for loop why are we iterating and what does the entire 'for block' does is important to be written.

I have added the comment `// Verify erasureReadFile() for random offsets and lengths.` but it is not adding value.

**Comment by krishnasrinivas** [comment]
> @krishnasrinivas : After the changes can you run this script with hash of your commit and the previous commit to see the performance changes https://github.com/minio/minio/blob/master/dist/benchmark/benchcmp.sh

@hackintoshrao the benchmark tool slowed down my laptop and gave fatal error: runtime: out of memory, what we can do is run it every night in cron and see if some benchmark has degraded, it can send a status mail and we can check which commit is causing the problem

```
krishna@escape:~/gocode/src/github.com/minio/minio$ dist/benchmark/benchcmp.sh  1229cfcaf0f5054f962ed0a122e74099e5d99678 cef26fd6eac29cbaa436eabb91178a679429068c
fetching Benchcmp...
Done.
Note: checking out '1229cfcaf0f5054f962ed0a122e74099e5d99678'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at 1229cfc... review fixes - 2
Running benchmarks at 1229cfcaf0f5054f962ed0a122e74099e5d99678
PASS
BenchmarkGetObjectVerySmallFS-4            30000         41120 ns/op        2296 B/op         28 allocs/op
BenchmarkGetObjectVerySmallXL-4             2000        613454 ns/op       84898 B/op       1217 allocs/op
BenchmarkGetObject10KbFS-4                 50000         40222 ns/op       23272 B/op         29 allocs/op
BenchmarkGetObject10KbXL-4                  2000        646717 ns/op      115626 B/op       1220 allocs/op
BenchmarkGetObject100KbFS-4                10000        144281 ns/op      215278 B/op         29 allocs/op
BenchmarkGetObject100KbXL-4                 2000        721639 ns/op      345013 B/op       1223 allocs/op
BenchmarkGetObject1MbFS-4                   1000       1509074 ns/op     3548971 B/op        127 allocs/op
BenchmarkGetObject1MbXL-4                   1000       1625628 ns/op     4277414 B/op       1227 allocs/op
BenchmarkGetObject5MbFS-4                    300       5733196 ns/op    15904924 B/op        543 allocs/op
BenchmarkGetObject5MbXL-4                    300       4723428 ns/op    16860321 B/op       1229 allocs/op
BenchmarkGetObject10MbFS-4                   100      11734373 ns/op    32595052 B/op       1063 allocs/op
BenchmarkGetObject10MbXL-4                   200      10154498 ns/op    33637535 B/op       1230 allocs/op
BenchmarkGetObject25MbFS-4                    50      28408166 ns/op    66150409 B/op       2623 allocs/op
BenchmarkGetObject25MbXL-4                    50      31602884 ns/op    67191969 B/op       1231 allocs/op
BenchmarkGetObject50MbFS-4                    30      47533779 ns/op    133348200 B/op      5224 allocs/op
BenchmarkGetObject50MbXL-4                    30      80656844 ns/op    134301501 B/op      1234 allocs/op
BenchmarkGetObject100MbFS-4                    2     683403990 ns/op    267874848 B/op     10425 allocs/op
BenchmarkGetObject100MbXL-4                   20     463548095 ns/op    268518637 B/op      1234 allocs/op
BenchmarkGetObject200MbFS-4                    2     756995547 ns/op    537059232 B/op     20826 allocs/op
BenchmarkGetObject200MbXL-4                    1    1167059376 ns/op    536954064 B/op      1234 allocs/op
BenchmarkGetObject500MbFS-4             fatal error: runtime: out of memory

runtime stack:
runtime.throw(0xcdf2f0, 0x16)
    /home/krishna/go/src/runtime/panic.go:530 +0x90
runtime.sysMap(0xcb2d130000, 0x1f400000, 0x433d00, 0x1104f58)
    /home/krishna/go/src/runtime/mem_linux.go:206 +0x9b
runtime.(*mheap).sysAlloc(0x10e7cc0, 0x1f400000, 0xc820063980)
    /home/krishna/go/src/runtime/malloc.go:429 +0x191
runtime.(*mheap).grow(0x10e7cc0, 0xfa00, 0x0)
    /home/krishna/go/src/runtime/mheap.go:651 +0x63
runtime.(*mheap).allocSpanLocked(0x10e7cc0, 0xfa00, 0x40f034)
    /home/krishna/go/src/runtime/mheap.go:553 +0x4f6
runtime.(*mheap).alloc_m(0x10e7cc0, 0xfa00, 0x100000000, 0x1010e90e0)
    /home/krishna/go/src/runtime/mheap.go:437 +0x119
runtime.(*mheap).alloc.func1()
    /home/krishna/go/src/runtime/mheap.go:502 +0x41
runtime.systemstack(0x7ff5cdf83de8)
    /home/krishna/go/src/runtime/asm_amd64.s:307 +0xab
runtime.(*mheap).alloc(0x10e7cc0, 0xfa00, 0x10100000000, 0xc82001b500)
    /home/krishna/go/src/runtime/mheap.go:503 +0x63
runtime.largeAlloc(0x1f400000, 0xc800000001, 0x4547b0)
    /home/krishna/go/src/runtime/malloc.go:766 +0xb3
runtime.mallocgc.func3()
    /home/krishna/go/src/runtime/malloc.go:664 +0x33
runtime.systemstack(0xc82017c640)
    /home/krishna/go/src/runtime/asm_amd64.s:291 +0x79
runtime.mstart()
    /home/krishna/go/src/runtime/proc.go:1048
```

**Comment by harshavardhana** [comment]
> BenchmarkGetObjectVerySmallFS-4            30000         41120 ns/op        2296 B/op         28 allocs/op
> BenchmarkGetObjectVerySmallXL-4             2000        613454 ns/op       84898 B/op       1217 allocs/op
> BenchmarkGetObject10KbFS-4                 50000         40222 ns/op       23272 B/op         29 allocs/op
> BenchmarkGetObject10KbXL-4                  2000        646717 ns/op      115626 B/op       1220 allocs/op
> BenchmarkGetObject100KbFS-4                10000        144281 ns/op      215278 B/op         29 allocs/op
> BenchmarkGetObject100KbXL-4                 2000        721639 ns/op      345013 B/op       1223 allocs/op
> BenchmarkGetObject1MbFS-4                   1000       1509074 ns/op     3548971 B/op        127 allocs/op
> BenchmarkGetObject1MbXL-4                   1000       1625628 ns/op     4277414 B/op       1227 allocs/op
> BenchmarkGetObject5MbFS-4                    300       5733196 ns/op    15904924 B/op        543 allocs/op
> BenchmarkGetObject5MbXL-4                    300       4723428 ns/op    16860321 B/op       1229 allocs/op
> BenchmarkGetObject10MbFS-4                   100      11734373 ns/op    32595052 B/op       1063 allocs/op
> BenchmarkGetObject10MbXL-4                   200      10154498 ns/op    33637535 B/op       1230 allocs/op
> BenchmarkGetObject25MbFS-4                    50      28408166 ns/op    66150409 B/op       2623 allocs/op
> BenchmarkGetObject25MbXL-4                    50      31602884 ns/op    67191969 B/op       1231 allocs/op
> BenchmarkGetObject50MbFS-4                    30      47533779 ns/op    133348200 B/op      5224 allocs/op
> BenchmarkGetObject50MbXL-4                    30      80656844 ns/op    134301501 B/op      1234 allocs/op
> BenchmarkGetObject100MbFS-4                    2     683403990 ns/op    267874848 B/op     10425 allocs/op
> BenchmarkGetObject100MbXL-4                   20     463548095 ns/op    268518637 B/op      1234 allocs/op
> BenchmarkGetObject200MbFS-4                    2     756995547 ns/op    537059232 B/op     20826 allocs/op
> BenchmarkGetObject200MbXL-4                    1    1167059376 ns/op    536954064 B/op      1234 allocs/op
> BenchmarkGetObject500MbFS-4             fatal error: runtime: out of memory
> 
> runtime stack:
> runtime.throw(0xcdf2f0, 0x16)
>     /home/krishna/go/src/runtime/panic.go:530 +0x90
> runtime.sysMap(0xcb2d130000, 0x1f400000, 0x433d00, 0x1104f58)
>     /home/krishna/go/src/runtime/mem_linux.go:206 +0x9b
> runtime.(_mheap).sysAlloc(0x10e7cc0, 0x1f400000, 0xc820063980)
>     /home/krishna/go/src/runtime/malloc.go:429 +0x191
> runtime.(_mheap).grow(0x10e7cc0, 0xfa00, 0x0)
>     /home/krishna/go/src/runtime/mheap.go:651 +0x63
> runtime.(_mheap).allocSpanLocked(0x10e7cc0, 0xfa00, 0x40f034)
>     /home/krishna/go/src/runtime/mheap.go:553 +0x4f6
> runtime.(_mheap).alloc_m(0x10e7cc0, 0xfa00, 0x100000000, 0x1010e90e0)
>     /home/krishna/go/src/runtime/mheap.go:437 +0x119
> runtime.(_mheap).alloc.func1()
>     /home/krishna/go/src/runtime/mheap.go:502 +0x41
> runtime.systemstack(0x7ff5cdf83de8)
>     /home/krishna/go/src/runtime/asm_amd64.s:307 +0xab
> runtime.(_mheap).alloc(0x10e7cc0, 0xfa00, 0x10100000000, 0xc82001b500)
>     /home/krishna/go/src/runtime/mheap.go:503 +0x63
> runtime.largeAlloc(0x1f400000, 0xc800000001, 0x4547b0)
>     /home/krishna/go/src/runtime/malloc.go:766 +0xb3
> runtime.mallocgc.func3()
>     /home/krishna/go/src/runtime/malloc.go:664 +0x33
> runtime.systemstack(0xc82017c640)
>     /home/krishna/go/src/runtime/asm_amd64.s:291 +0x79
> runtime.mstart()
>     /home/krishna/go/src/runtime/proc.go:1048

We need to relook at why our benchmarks try to malloc so much, 'mallocs' are not cheap on Golang we should optimize this part.
